### PR TITLE
Follow up to universal System.Linq.Expressions

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Most of this is test changes so that we run more tests when the expression IL compiler is disabled. There is one product change that we don't need to take, but felt like an improvement. I'll put comments inline in a bit.

### CODE REVIEW

File: src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/DelegateHelpers.cs [hunk 1]
```diff
@@ -109,7 +109,7 @@ private static bool IsByRef(DynamicMetaObject mo)
 
         private static Type MakeNewCustomDelegate(Type[] types)
         {
-            if (LambdaExpression.CanCompileToIL)
+            if (RuntimeFeature.IsDynamicCodeSupported)
```
Comment (MichalStrehovsky) [inline src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/DelegateHelpers.cs:112]: PRODUCT CHANGE: the existing setting was mechanically propagated from FEATURE_COMPILE, but I think here we should be more nuanced between "IL expression compiler is disabled" and "the platform supports reflection emit".

This codepath is not related to compiling IL expressions and we can allow it if emit is otherwise supported. It doesn't feel like it should be lumped into IL expression compiler not being enabled.

If we don't take this change, we need to disable a couple more tests on IL compiler being unavailable.

File: src/libraries/System.Linq.Expressions/tests/Dynamic/InvokeMemberBindingTests.cs [hunk 2]
```diff
@@ -54,7 +54,7 @@ public static IEnumerable<object[]> ObjectArguments()
             yield return new[] {new object()};
         }
 
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/55070", typeof(PlatformDetection), nameof(PlatformDetection.IsLinqExpressionsBuiltWithIsInterpretingOnly))]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/55070", TestPlatforms.iOS | TestPlatforms.tvOS | TestPlatforms.MacCatalyst)]
```
Comment (MichalStrehovsky) [inline src/libraries/System.Linq.Expressions/tests/Dynamic/InvokeMemberBindingTests.cs:57]: This test is passing with CoreCLR and IL compiler disabled, so whatever this is, it's iDevice-specific. This test was disabled in #54970 and not part of the mechanical refactoring.

File: src/libraries/System.Linq.Expressions/tests/Dynamic/InvokeMemberBindingTests.cs [hunk 3]
```diff
@@ -335,8 +331,7 @@ private class ManyOverloads
                 int arg10) => 11;
         }
 
-        // We're not testing compilation, but we do need Reflection.Emit for the test
-        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotLinqExpressionsBuiltWithIsInterpretingOnly))]
+        [Fact]
```
Comment (MichalStrehovsky) [inline src/libraries/System.Linq.Expressions/tests/Dynamic/InvokeMemberBindingTests.cs:334]: This was probably hitting some .NET Native specific issues with dynamic keyword. I don't see it using Ref.Emit anywhere so leaving like this.

File: src/libraries/System.Linq.Expressions/tests/Dynamic/InvokeMemberBindingTests.cs [hunk 4]
```diff
@@ -380,8 +375,7 @@ public static IEnumerable<object[]> SameNameObjectPairs()
             return testObjects.SelectMany(i => testObjects.Select(j => new[] { i, j }));
         }
 
-        // We're not testing compilation, but we do need Reflection.Emit for the test
-        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsNotLinqExpressionsBuiltWithIsInterpretingOnly))]
+        [Theory]
```
Comment (MichalStrehovsky) [inline src/libraries/System.Linq.Expressions/tests/Dynamic/InvokeMemberBindingTests.cs:378]: Also don't see anything ref.emitting. Was probably some .NET Native issue related to dynamic.

File: src/libraries/System.Linq.Expressions/tests/IndexExpression/IndexExpressionTests.cs [hunk 5]
```diff
@@ -464,15 +464,15 @@ public void IndexedPropertyVarArgs()
             AssertExtensions.Throws<ArgumentException>("propertyName", () => Expression.Property(instance, "Item", Expression.Constant(0), Expression.Constant(0), Expression.Constant(0)));
         }
 
-        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotLinqExpressionsBuiltWithIsInterpretingOnly))]
+        [Fact]
```
Comment (MichalStrehovsky) [inline src/libraries/System.Linq.Expressions/tests/IndexExpression/IndexExpressionTests.cs:467]: Don't see ref emitting here and it passes with IL expression compiler disabled, so leaving like this.

File: src/libraries/System.Linq.Expressions/tests/InterpreterTests.cs [hunk 6]
```diff
@@ -101,9 +101,7 @@ .maxcontinuation 1
                   }");
         }
 
-        // IsNotLinqExpressionsBuiltWithIsInterpretingOnly is not directly required,
-        // but this functionality relies on private reflection and that would not work with AOT
-        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotLinqExpressionsBuiltWithIsInterpretingOnly))]
+        [Fact]
```
Comment (MichalStrehovsky) [inline src/libraries/System.Linq.Expressions/tests/InterpreterTests.cs:104]: Original motivation sounds .NET Native related.

File: src/libraries/System.Linq.Expressions/tests/New/NewWithByRefParameterTests.cs [hunk 7]
```diff
@@ -70,7 +70,8 @@ public void CreateByRefAliasingInterpreted()
             CreateByRefAliasing(useInterpreter: true);
         }
 
-        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotLinqExpressionsBuiltWithIsInterpretingOnly))]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/19286", typeof(PlatformDetection), nameof(PlatformDetection.IsLinqExpressionsBuiltWithIsInterpretingOnly))]
```
Comment (MichalStrehovsky) [inline src/libraries/System.Linq.Expressions/tests/New/NewWithByRefParameterTests.cs:73]: Switched this one to ActiveIssue - the preceding test has the motivation for this.

**Review by MichalStrehovsky — COMMENTED** [review]
I've validated that we run the same number of tests in default CoreCLR configuration.

I've also validated that the test fully passes when I artificially disable IL expression compiler on CoreCLR.

**Review by stephentoub — APPROVED** [review]

**Comment by MichalStrehovsky** [comment]
@cston do you have any objection to the change in DelegateHelpers.cs?

The behavior change is basically:

1. In deployment models with expression compiler disabled, but ref emit available, we would newly not throw PNSE, but make things work.
2. In deployment models with expression compiler disabled, but ref emit unavailable, no change.
3. In deployment models with ref emit unavilable, PNSE, but from a different codepath (Ref.Emit would throw originally, now we short circuit earlier).
4. In deployment models with expression compiler enabled and ref.emit available, no change.

Basically, it either makes things work, which sounds like a good thing, or just throws PNSE from a different place.

**Review by cston — APPROVED** [review]
